File: haskell/difference-of-squares/src/Squares.hs
module Squares (difference, squareOfSums, sumOfSquares) where

difference :: Integral a => a -> a
difference n = squareOfSums n - sumOfSquares n

squareOfSums :: Integral a => a -> a
squareOfSums n = square $ sum [0..n]

sumOfSquares :: Integral a => a -> a
sumOfSquares n = sum $ map square [0..n]

square :: Integral a => a -> a
square n = n * n

File: ruby/difference-of-squares/squares.rb
require 'delegate'

class MyNum < DelegateClass(Fixnum)
  def square
    MyNum.new self * self
  end

  def +(number)
    MyNum.new self.to_i + number
  end
end

class Squares
  attr_reader :number

  def initialize(number)
    @number = number
  end

  def square_of_sums
    range.reduce(&:+).square
  end

  def sum_of_squares
    range.map(&:square).reduce(&:+)
  end

  def difference
    square_of_sums - sum_of_squares
  end

  private

  def range
    (1..number).map { |n| MyNum.new n }
  end
end
